### ISSUE TITLE
Question about handicap

### ISSUE THREAD

**Issue by AnddyAnddy** [haxball/haxball-issues#1872]
https://i.imgur.com/nu9POP3.mp4
In this video, I am watching 2 players:
- Cèdres, playing with handicap 200
- Afa, playing with handicap 0

And my point of view is around 100 ping with extrapolation and handicap 0

We can see that the player with 200 handicap is MUCH smoother than the player with 0 handicap.
I don't know how exactly the handicap is working but here goes my question:
Could the client (my POV) act like every other player in the room is having a 200 handicap ? Even though it is not really true regarding the actual game state

**Comment by basro** [comment]
I'll attempt to explain what handicap does.

Handicap can be understood as extrapolation + equal amounts of increased input lag.

If you were to remove the extrapolation component when playing with handicap 200 you'd only see the effect of your actions 200msec after you pressed a key.

While your action is delayed, it is notified to everyone else immediately. This helps everyone else predict your actions better. If someone has 200msec ping but you set your handicap to 200 then they will not see you warping at all.

**Comment by gokco7** [comment]
> 
But what do you mean by saying "warping"? Could you make a 5-10 seconds video to show it properly? 
And, still I have some tiny confusions about handicap, should everyone in the room use handicap at least a little? to make it more efficient, because I think handicap is not usefull at all

By the way, thank you for the time you are spending to improve the game and being open to talk with the players after a long and long time

Regards

**Comment by basro** [comment]
With warping I mean big changes to the predicted state of the game. Setting extrapolation to a high value causes warping on every one of your key presses.

Warping happens when the game's prediction differed from what really happened.

Handicap makes other people see you less 'warpy' but it makes you see other people with more warping by the same amount. This means that if two players set the same handicap value they will compensate the effects and see each other with the same amount of warping. However other observers that did not set any handicap will see these two players with less warp.

Therefore, there's no point in everyone in the room having handicap.

The only real use case for this feature is to balance the playing field when some players have more ping than others.

Say one player in a room has 20ms ping and everyone else has 120ms ping. If the player with 20msec ping sets his handicap to 50 (which is (120-20) / 2) he will see everyone else as if his real ping was 60 (20+60), and everyone else will see **_him_** move as if they themselves had 60 ping. Of course this is not 100% balancing out the fairness because while everyone else will see him less warpy they will still see the other non handicapped players warp the full 120ms distance.

**Comment by kingjasko** [comment]
Do you see it possible to give rooms the ability to somehow sync everyone to the same ping+handicap? Example with 2 players, 1 with 20 ping and other one with 30 ping, the room could automatically detect their ping (perhaps it could be run every tick if not too demanding) and add handicap to match the desired number, so it could let's say add 30 and 20 handicap to the 2 players, so everyone plays with 50 handicap+ping.

If you see it being possible, it might be a decent feature in the future.

**Comment by gokco7** [comment]
> With warping I mean big changes to the predicted state of the game. Setting extrapolation to a high value causes warping on every one of your key presses.
> 
> Warping happens when the game's prediction differed from what really happened.
> 
> Handicap makes other people see you less 'warpy' but it makes you see other people with more warping by the same amount. This means that if two players set the same handicap value they will compensate the effects and see each other with the same amount of warping. However other observers that did not set any handicap will see these two players with less warp.
> 
> Therefore, there's no point in everyone in the room having handicap.
> 
> The only real use case for this feature is to balance the playing field when some players have more ping than others.
> 
> Say one player in a room has 20ms ping and everyone else has 120ms ping. If the player with 20msec ping sets his handicap to 50 (which is (120-20) / 2) he will see everyone else as if his real ping was 60 (20+60), and everyone else will see **_him_** move as if they themselves had 60 ping. Of course this is not 100% balancing out the fairness because while everyone else will see him less warpy they will still see the other non handicapped players warp the full 120ms distance.

Alright, I got it, thanks for explaining detailed

**Comment by punisherhaxball** [comment]
But I have one question how does Haxball gameplay is so related with Ping? I mean why other players moving warpy and ball animation starts late when we have high ping? CSGO, Rocket League, Valorant all these games don’t have that problem. Of course in these games you feel the heaviness comes with high ping but that doesn’t change how game appears in your screen. I think that’s a really big problem in this game and you can solve that. 💪🏻

Easy to reproduce just see the problem yourself. Set your extrapolation to 0, then play in a room where you have 10 ping, after then join a room where you get 70 ping and just compare visuals(animations). Just the animations, for example compare how ball starts its animation when others shoot it when ping is low and high, you’ll find the answer yourself. ✌🏻

Please don’t ignore this message even a technical comment will be so much appreciated by us.👋🏻🙏🏻

**Comment by punisherhaxball** [comment]
And is there a possibility to have extrapolate game only for our player? For example, I want to see my player warpy but not the others, I just don’t want to predict their movement, all I care is my player. So why not make a new extrapolation like that? You created extrapolation feature I think you can also choose which things it should predict and which not.

**Comment by basro** [comment]
Rocket league uses more or less the same prediction system as HaxBall does and it suffers the same warp effect from mispredictions. Unlike HaxBall, Rocket League has a little bit of interpolation applied to the ball and opponent cars which means you don't see things outright teleport to a new location when there's a misprediction, instead you see them relocate very fast. I'm not a fan of interpolation since it adds a bit of extra time before your eyes are seeing the new correctly predicted physical state.

Lag compensation in CSGO and Valorant is different, but it's a system that works well for hitscan FPS but is not useful for actual physical interactions between objects.


> And is there a possibility to have extrapolate game only for our player?

It's technically 'doable' but it would be really annoying to implement and I think it's too weird of a feature. You'd be seeing yourself clipping inside all the other physical objects. For example when pushing the ball you'd see yourself inside of the ball (or even past it) before the ball even moves.

**Comment by basro** [comment]
@kingjasko 

> Do you see it possible to give rooms the ability to somehow sync everyone to the same ping+handicap? Example with 2 players, 1 with 20 ping and other one with 30 ping, the room could automatically detect their ping (perhaps it could be run every tick if not too demanding) and add handicap to match the desired number, so it could let's say add 30 and 20 handicap to the 2 players, so everyone plays with 50 handicap+ping.
> 
> If you see it being possible, it might be a decent feature in the future.

It is possible to implement, although kinda tricky to do well.

I'm not sure it's the best idea though, I don't think that players with low ping should be penalized because a player with high ping has joined the room.

High ping being a problem for the high ping player creates the correct incentives: Players wanting to join rooms near them and wanting to have a properly working internet connection.

**Comment by AnddyAnddy** [comment]
What are your recommandations about handicap on concrete cases
Let's say we are in 4v4, 5 players have less than 40 ping, 1 has 70 and the 2 others have 100. Let's pretend everyone has extrapolation 0

Is there any way we can make this less laggy for the high pings and not that much more laggy for the low pings ?

**Comment by punisherhaxball** [comment]
I agree with the feature of handicapping everyone to some fixed value with a command in the room. But it should be really unbypassable LIMITED. For example if a player have 80 ping and other players have 20 30 40. When that feature is enabled, the room should sync everyone's handicap to highest player's ping.  

For example: 
Player Z: **80 ping** + no forced handicap 
Player A: 20 real ping + 60 forced handicap
Player B: 30 real ping + 50 forced handicap
Player C: 40 real ping + 40 forced handicap

And if the player have highest ping in the room the handicap wouldn't effect him because there is no need make the highest players game play worse. 

I think this feature could be really cool because not everyone sitting next to Hetzner or OVH or Google Cloud servers and many people don't like to play while having unfair pings.

**Comment by basro** [comment]
> What are your recommandations about handicap on concrete cases
> Let's say we are in 4v4, 5 players have less than 40 ping, 1 has 70 and the 2 others have 100. Let's pretend everyone has extrapolation 0
> Is there any way we can make this less laggy for the high pings and not that much more laggy for the low pings ?

In that situation I'd tell everyone to set their handicap to (100 - player_ping) / 2.

The players with ping 40 will set handicap 30.
The player with ping 70 will set handicap 15.
The players with ping 100 will set handicap 0.

This will have the following effect:

**The players with ping 40**
* will see ping 40 players with the same amount of warpyness as before. Total warping perceived 40ms.
* will see ping 70 players with 15ms of extra warping. Total warping perceived 55ms.
* will see ping 100 players with 30ms of extra warping. Total warping perceived 70ms.

**The players with ping 70**
* will see ping 40 players with 15ms less warping as before. Total warping perceived 55ms.
* will see ping 70 players with the same amount of warpyness as before. Total warping perceived 70ms.
* will see ping 100 players with 15ms of extra warping. Total warping preceived 85ms.

**The players with ping 100**
* will see ping 40 players with 30ms less warping as before. Total warping perceived 70ms.
* will see ping 70 players with 15ms less warping as before. Total warping perceived 85ms.
* will see ping 100 players with the same amount of warping as before. Total warping preceived 100ms.

**Comment by kingjasko** [comment]
If I remember correctly, wasn't handicap visible in the lobby for everyone to see in the past? I don't see that anymore, either way I think it'd be nice to be able to see everyone's handicap AND extrapolation publicly in the lobby and in replays as well (perhaps with a toggle button in settings).


> I'm not sure it's the best idea though, I don't think that players with low ping should be penalized because a player with high ping has joined the room.

It's not about penalization, it'd give hosters and communities a chance to create an even fairer competition between us, especially for a continent like Europe, with good connectivity, I had an idea for a long time to create a competition where everyone is playing with minimum 50 handicap+ping, anyone over that would just play the same way to not create too many issues for people with lower pings. Unlike in other games, haxball is very touchy when it comes to ping differences, even a 5-10ms change you can absolutely feel if you play this game often.

> High ping being a problem for the high ping player creates the correct incentives: Players wanting to join rooms near them and wanting to have a properly working internet connection.

I generally agree with that, it would remain the same for the most part, just that it'd give another option for those who choose to use it.

I think both publicly visible handicap and extrapolation of players along with some sort of room's automation when it comes to calculating and setting handicap based on player's ping would be quite nice, but the latter shouldn't be a priority just yet.

**Comment by basro** [comment]
> If I remember correctly, wasn't handicap visible in the lobby for everyone to see in the past? I don't see that anymore, either way I think it'd be nice to be able to see everyone's handicap AND extrapolation publicly in the lobby and in replays as well (perhaps with a toggle button in settings).

Yes this was a feature in flash but was not replicated in html5.



> It's not about penalization, it'd give hosters and communities a chance to create an even fairer competition between us, especially for a continent like Europe, with good connectivity, I had an idea for a long time to create a competition where everyone is playing with minimum 50 handicap+ping, anyone over that would just play the same way to not create too many issues for people with lower pings. Unlike in other games, haxball is very touchy when it comes to ping differences, even a 5-10ms change you can absolutely feel if you play this game often.

Yeah, for that use case it makes a lot of sense. I'll definitely consider it for the future.

**Comment by punisherhaxball** [comment]
Please make it happen. All league organizations can have fair matches.
